proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,381 @@
+types. Importantly though those new methods cannot be back-deployed since they
+introduce new types. Those are the new typed throw APIs that we propose to add:
+
+```swift
+extension AsyncSequence {
+  public func map<Transformed, Failure: Error>(
+    _ transform: @Sendable @escaping (Element) async throws -> Transformed
+  ) -> some AsyncSequence<ElementOfResult, Failure>
+
+  public func compactMap<ElementOfResult, Failure: Error>(
+    _ transform: @Sendable @escaping (Element) async throws(Failure) -> ElementOfResult?
+  ) -> some AsyncSequence<ElementOfResult, Failure>
+  
+  public func drop(
+    while predicate: @Sendable @escaping (Element) async throws(Failure) -> Bool
+  ) -> some AsyncSequence<ElementOfResult, Failure>
+
+  public func filter(
+    _ isIncluded: @Sendable @escaping (Element) async throws(Failure) -> Bool
+  ) -> some AsyncSequence<ElementOfResult, Failure>
+
+  public func flatMap<SegmentOfResult: AsyncSequence>(
+    _ transform: @Sendable @escaping (Element) async throws(Failure) -> SegmentOfResult
+  ) -> some AsyncSequence<ElementOfResult, Failure>
+
+  public func prefix(
+    while predicate: @Sendable @escaping (Element) async throws(Failure) -> Bool
+  ) -> some AsyncSequence<ElementOfResult, Failure> // This is currently rethrows for no reason: https://github.com/apple/swift/issues/66922
+}
+```
+
+Open question:
+- I am unsure if we can actually adopt opaque return types here due to the `Sendable` constraints. By default
+those do not propagate through opaque return types. We could overload those methods to conditionally return `some (AsyncSequence<ElementOfResult, Failure> & Sendable)`. 
+- The `@Sendable` on the various closures feels overly restrictive and once we solve the hopping of `next` it shouldn't be required. In line with the previous
+open question we probably want to make the sendability of the return type be conditional on if the closure is `Sendable`
+
+## Source compatibility
+
+TODO
+
+## Effect on ABI stability
+
+The ABI between an function with an untyped throws and one that uses typed
+throws will be different, so that typed throws can benefit from knowing the
+precise type. For most of the proposed changes, an actual ABI break can
+be avoided because the implementations can make use of
+[`@backDeploy`](https://github.com/apple/swift-evolution/blob/main/proposals/0376-function-back-deployment.md).
+However, the suggested change to `AsyncIteratorProtocol` might not be able to be
+made in a manner that does not break ABI stability.
+
+## Effect on API resilience
+
+TODO
+
+## Future directions
+
+TODO
+
+## Alternatives considered
+
+### Adopt typed throws for `Clock.sleep`
+
+Most of the clock implementation only throw a `CancellationError` from their
+`sleep` method; however, nothing enforces this right now and there might be
+implementations out there that throw a different error. Restricting the protocol
+to only throw `CancellationError`s would be a breaking change
```

Comment:
IIRC it was part of the design consideration that folks could throw whatever error type they wanted.